### PR TITLE
libsbol 2.1 (new formula)

### PR DESCRIPTION
**Pull request by bbartley**
Hello,

This is a new formula for libSBOL, a C++ library that supports computer-aided design for synthetic biology and file I/O in the SBOL file format.

### Have you:

- [X] Followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-science/blob/master/.github/CONTRIBUTING.md) document?
- [X] Checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-science/pulls) for the same update/change?  
- [X] Run `brew audit --strict --online <formula>` (where `<formula>` is the name of the formula you're submitting) and corrected all errors?
- [X] Built your formula locally prior to submission with `brew install <formula>`?

### New formula: have you

- [X] Written a sensible test? (The best test is to compile and run a sample program.)

### Updates to existing formula: have you

- [ ] Removed the `revision` line, if any, when bumping the version number?
- [ ] Added/bumped the `revision` number if the changes affect the precompiled bottles?
- [ ] Not altered the `bottle` section?
- [ ] Checked that the tests still pass?

### CODE REVIEW

**Comment by jonchang** [comment]
Hi @bbartley, thanks for contributing to Homebrew! It looks like you'll have to help cmake find some libraries. Feel free to `depends_on "curl" if OS.linux?` or similar, but fixing the macOS errors is most important.

macOS fails with:

```
-- Could NOT find PkgConfig (missing:  PKG_CONFIG_EXECUTABLE) 
-- Could NOT find Raptor (missing:  RAPTOR_LIBRARY RAPTOR_INCLUDE_DIR) 
-- Found CURL: /usr/lib/libcurl.dylib  
-- Found LibXslt: /usr/lib/libxslt.dylib  
Found raptor2 library at 
Copying /tmp/libsbol-20161213-75050-zcczmk/libSBOL-2.1.0/source/raptor/include/raptor2/raptor.h;/tmp/libsbol-20161213-75050-zcczmk/libSBOL-2.1.0/source/raptor/include/raptor2/raptor2.h;/tmp/libsbol-20161213-75050-zcczmk/libSBOL-2.1.0/source/raptor/include/raptor2/raptor_internal.h to /tmp/libsbol-20161213-75050-zcczmk/libSBOL-2.1.0/release/library/sbol
Configuring for x64
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
RAPTOR_INCLUDE_DIR (ADVANCED)
   used as include directory in directory /tmp/libsbol-20161213-75050-zcczmk/libSBOL-2.1.0/source
   used as include directory in directory /tmp/libsbol-20161213-75050-zcczmk/libSBOL-2.1.0/source
   used as include directory in directory /tmp/libsbol-20161213-75050-zcczmk/libSBOL-2.1.0/source
   used as include directory in directory /tmp/libsbol-20161213-75050-zcczmk/libSBOL-2.1.0/source
   used as include directory in directory /tmp/libsbol-20161213-75050-zcczmk/libSBOL-2.1.0/source
   used as include directory in directory /tmp/libsbol-20161213-75050-zcczmk/libSBOL-2.1.0/source
   used as include directory in directory /tmp/libsbol-20161213-75050-zcczmk/libSBOL-2.1.0/source
RAPTOR_LIBRARY (ADVANCED)
    linked by target "sbol64" in directory /tmp/libsbol-20161213-75050-zcczmk/libSBOL-2.1.0/source
```

Linux fails with:

```
-- Could NOT find Raptor (missing:  RAPTOR_LIBRARY RAPTOR_INCLUDE_DIR) 
-- Could NOT find CURL (missing:  CURL_LIBRARY) 
LibXslt: LIBXSLT_LIBRARIES-NOTFOUND
-- Could NOT find LibXslt (missing:  LIBXSLT_LIBRARIES LIBXSLT_INCLUDE_DIR) 
Found raptor2 library at 
Copying /tmp/libsbol-20161213-38333-z953to/libSBOL-2.1.0/source/raptor/include/raptor2/raptor.h;/tmp/libsbol-20161213-38333-z953to/libSBOL-2.1.0/source/raptor/include/raptor2/raptor2.h;/tmp/libsbol-20161213-38333-z953to/libSBOL-2.1.0/source/raptor/include/raptor2/raptor_internal.h to /tmp/libsbol-20161213-38333-z953to/libSBOL-2.1.0/release/library/sbol
Configuring for x64
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
CURL_LIBRARY (ADVANCED)
    linked by target "sbol64" in directory /tmp/libsbol-20161213-38333-z953to/libSBOL-2.1.0/source
LIBXSLT_LIBRARIES (ADVANCED)
    linked by target "sbol64" in directory /tmp/libsbol-20161213-38333-z953to/libSBOL-2.1.0/source
RAPTOR_INCLUDE_DIR (ADVANCED)
   used as include directory in directory /tmp/libsbol-20161213-38333-z953to/libSBOL-2.1.0/source
   used as include directory in directory /tmp/libsbol-20161213-38333-z953to/libSBOL-2.1.0/source
   used as include directory in directory /tmp/libsbol-20161213-38333-z953to/libSBOL-2.1.0/source
   used as include directory in directory /tmp/libsbol-20161213-38333-z953to/libSBOL-2.1.0/source
   used as include directory in directory /tmp/libsbol-20161213-38333-z953to/libSBOL-2.1.0/source
   used as include directory in directory /tmp/libsbol-20161213-38333-z953to/libSBOL-2.1.0/source
   used as include directory in directory /tmp/libsbol-20161213-38333-z953to/libSBOL-2.1.0/source
RAPTOR_LIBRARY (ADVANCED)
    linked by target "sbol64" in directory /tmp/libsbol-20161213-38333-z953to/libSBOL-2.1.0/source
```

**Comment by bbartley** [comment]
Hi @jonchang , I've added dependencies to the formula and hopefully resolved the build error.  Thanks for the help, and happy holidays!

**Comment by bbartley** [comment]
@jonchang looks like I still have a few integration tests to pass, like missing compiler flags.  I was hoping to get this pulled by Christmas, but it looks like I still have a few things to figure out!  No worries, I'm starting to get the hang of it.

**Comment by jonchang** [comment]
No worries, ping me if you need assistance.

**Comment by bbartley** [comment]
Hi @jonchang, CI test https://travis-ci.org/Homebrew/homebrew-science/builds/186843491 for libSBOL fails on line 1625 because CMake doesn't know the correct path for dependencies which homebrew has installed.  Any hint where should I tell CMake to look?  On my Mac machine, it looks in /usr/local/lib and /usr/local/include for libraries installed via homebrew.  Is the install path different for Linux?  Thanks for the help!

**Comment by jonchang** [comment]
@bbartley thanks for working on this and sorry for the late response. You probably have to manually specify the path to the dependencies; try doing `cd $(brew --repo homebrew/core) && grep _INCLUDE_DIR -r .` for some examples (e.g., `open-scene-graph` from core).

**Comment by jonchang** [comment]
@bbartley I got it to build on Linux like this. 

```ruby
class Sbol < Formula
  desc "Read and write files in the Synthetic Biology Open Language (SBOL)"
  homepage "https://synbiodex.github.io/libSBOL"
  url "https://github.com/SynBioDex/libSBOL/archive/linuxbrew-3.tar.gz"
  sha256 "e07cb3009d8e7252547e665a9eb5f2f5ca7750a0b2143a7dd237216a33b498c0"

  depends_on "cmake" => :build
  depends_on "pkg-config" => :build
  depends_on "jsoncpp"
  depends_on "curl"
  depends_on "libxslt"
  depends_on "raptor"

  def install
    raptor = Formula["raptor"]
    libxslt = Formula["libxslt"]
    jsoncpp = Formula["jsoncpp"]
    args = %W[
      -DSBOL_BUILD_SHARED=TRUE
      -DSBOL_BUILD_STATIC=FALSE
      -DRAPTOR_INCLUDE_DIR=#{raptor.opt_include}
      -DLIBXSLT_INCLUDE_DIR=#{libxslt.opt_include}
      -DJsonCpp_LIBRARY=#{jsoncpp.opt_prefix}/lib64
    ]
    cd "source" do
      %w[document.cpp document.h object.cpp object.h properties.h property.h].each do
        |f| inreplace f, "raptor2.h", "raptor2/raptor2.h"
      end
    end
    system "cmake", ".", *args, *std_cmake_args
    system "make", "install"
  end

  test do
    (testpath/"test.cpp").write <<-EOS.undent
      #include "sbol.h"

      using namespace sbol;

      int main() {
        Document& doc = *new Document();
        doc.write("test.xml");
        return 0;
      }
    EOS
    system ENV.cc, "test.cpp", "-o", "test", "-std=c++11", "-I#{HOMEBREW_PREFIX}/include/curl", "-I#{HOMEBREW_PREFIX}/include", "-I#{HOMEBREW_PREFIX}/include/raptor2", "-I#{include}/sbol", "-L#{HOMEBREW_PREFIX}/lib", "-L#{lib}", "-ljsoncpp", "-lcurl", "-lraptor2", "-lsbol", "-lc++"
    system "./test"
  end
end
```

**Comment by jonchang** [comment]
Note that `lib64` is probably `lib` on macOS (so you may want to add an `if OS.linux?` guard somewhere). Ping me if you need any help.

**Comment by jonchang** [comment]
As long as it works on macOS it's ok. Note that defining the include directories (in this comment https://github.com/Homebrew/homebrew-science/pull/4715#issuecomment-269885264 ) seems to get it to build on Linux so if you want to you can use that as a basis to open a new pull request.

Please remote the debugging code (`puts ENV.cxx`) and so forth from the test block and this should be ready to merge.

File: sbol.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+      using namespace sbol;
+
+      int main() {
+        Document& doc = *new Document();
+        doc.write("test.xml");
+        return 0;
+      }
+    EOS
+    if ENV["CC"]["gcc"]
```
Comment (jonchang) [inline sbol.rb]: Does `ENV.compiler == :gcc` work?

**Comment by jonchang** [comment]
Thanks for contributing to Homebrew @bbartley! This has been merged.

**Pull request closed without merging**